Implement TCP keepalive support from RFC 1122

maweicoltsmart opened the pull request:
Fixes zephyrproject-rtos#34044
 Signed-off-by:  Horse Ma <mawei@coltsmart.com>

jukkar commented on samples/net/sockets/echo_client/src/tcp.c:
```diff
@@ -158,7 +158,33 @@ static int start_tcp_proto(struct data *data, struct sockaddr *addr,
 #endif
-
+#if defined(CONFIG_NET_TCP_KEEPALIVE)
+	int keep_alive = true; // enable keep alive probe
+	int keep_idle = 5; // 5 Second
+	int keep_intvl = 5; // 5 Second
+	int keep_cnt = 3; // probe 3 times
```
Please use `/* */` comments

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -545,6 +545,9 @@ static void tcp_send_timer_cancel(struct tcp *conn)
 	k_work_cancel_delayable(&conn->send_timer);
+#if defined(CONFIG_NET_TCP_KEEPALIVE)
+	k_work_cancel_delayable(&conn->keepalive_timer);
+#endif /*CONFIG_NET_TCP_KEEPALIVE */
```
style nit, this looks better if `/* CONFIG....`, please do search and replace for the whole file

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -879,6 +882,132 @@ static int get_tcp_nodelay(struct tcp *conn, void *value, size_t *len)
+	}
+
+	keep_alive = *(int *)value;
+
```
we typically do not put empty line between assignment and check

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -879,6 +882,132 @@ static int get_tcp_nodelay(struct tcp *conn, void *value, size_t *len)
+	}
+
+	keep_idle = *(int *)value;
+
```
ditto

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -879,6 +882,132 @@ static int get_tcp_nodelay(struct tcp *conn, void *value, size_t *len)
+	}
+
+	keep_intvl = *(int *)value;
+
```
ditto

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -879,6 +882,132 @@ static int get_tcp_nodelay(struct tcp *conn, void *value, size_t *len)
+	}
+
+	keep_cnt = *(int *)value;
+
```
ditto

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -879,6 +882,132 @@ static int get_tcp_nodelay(struct tcp *conn, void *value, size_t *len)
+	if (len) {
+		*len = sizeof(int);
+	}
+	return 0;
```
also we usually put empty line between `}` and next statement

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -879,6 +882,132 @@ static int get_tcp_nodelay(struct tcp *conn, void *value, size_t *len)
+	if (len) {
+		*len = sizeof(int);
+	}
+	return 0;
```
ditto

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -879,6 +882,132 @@ static int get_tcp_nodelay(struct tcp *conn, void *value, size_t *len)
+	if (len) {
+		*len = sizeof(int);
+	}
+	return 0;
```
and here

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -879,6 +882,132 @@ static int get_tcp_nodelay(struct tcp *conn, void *value, size_t *len)
+	if (len) {
+		*len = sizeof(int);
+	}
+	return 0;
```
and here too

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -1283,6 +1414,76 @@ static void tcp_fin_timeout(struct k_work *work)
+
+	pkt = tcp_pkt_alloc(conn, sizeof(struct tcphdr));
+	if (!pkt) {
+		ret = -ENOBUFS;
```
this is no op as the function does not return anything

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -1283,6 +1414,76 @@ static void tcp_fin_timeout(struct k_work *work)
+	ret = ip_header_add(conn, pkt);
+	if (ret < 0) {
+		tcp_pkt_unref(pkt);
+		goto out;
```
perhaps nicer to add `unref:` label at the end and jump to it

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -1283,6 +1414,76 @@ static void tcp_fin_timeout(struct k_work *work)
+
+	ret = tcp_header_add(conn, pkt, ACK, conn->seq - 1);
+	if (ret < 0) {
+		tcp_pkt_unref(pkt);
+		goto out;
```
and here

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -1369,6 +1570,15 @@ static struct tcp *tcp_conn_alloc(struct net_context *context)
+	conn->keep_idle = CONFIG_NET_TCP_KEEPIDLE_DEFAULT;
+	conn->keep_intvl = CONFIG_NET_TCP_KEEPINTVL_DEFAULT;
+	conn->keep_cnt = CONFIG_NET_TCP_KEEPCNT_DEFAULT;
+	NET_INFO("keepalive timer init idle = %d, interval = %d, cnt = %d", conn->keep_idle, conn->keep_intvl, conn->keep_cnt);
```
Not sure if this needs to be in info level, would user be interested to see this always. I would suggest `NET_DBG` here instead.

jukkar reviewed (commented):
Looks good, minor style nits.

maweicoltsmart committed:
net/tcp: Implement Keepalive Support

Fixes #34044

When a TCP connection is established, if there is no data exchange between the two parties within the set time, the side that enables TCP Keepalive will send a TCP probe packet with the same sequence number as the previous tcp packet. This TCP probe packet is a pure ACK packet (the specification recommends that it should not contain any data, but can also contain 1 nonsense byte, such as 0x00.). If there is no response from the other side after several consecutive probe packets are sent, it is determined that the tcp connection has failed, and the connection is closed.

Signed-off-by: Horse Ma <mawei@coltsmart.com>

ssharks commented:
@maweicoltsmart Although the review process is intended to aid developers, I can imagine you are overwhelmed by the amount of review comments. 

Some tips you can used to check by yourself:

Run the most prominent tests for TCP:
./scripts/twister -T tests/net/socket/tcp -p qemu_x86
./scripts/twister -T tests/net/tcp -p qemu_x86

Check your head commit before pushing against a lot of style rules
./scripts/checkpatch.pl -g HEAD

I hope this helps you with this and hopefully many more PRs

maweicoltsmart commented on subsys/net/ip/tcp.c:
```diff
@@ -545,6 +545,9 @@ static void tcp_send_timer_cancel(struct tcp *conn)
 
 	k_work_cancel_delayable(&conn->send_timer);
+#if defined(CONFIG_NET_TCP_KEEPALIVE)
+	k_work_cancel_delayable(&conn->keepalive_timer);
```
Yes, this code is redundant and may cause keepalive to fail, I'll correct it right away.
